File: lib/primary/of55-1.js
/**
 * 二叉树的最大深度
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
export var maxDepth = function(root) {
    if (root === null) return 0;
    let [left, right] = [0, 0];
    left = maxDepth(root.left)
    right = maxDepth(root.right)
    return ( left > right ? left : right ) + 1
};

File: src/primary/of55_1.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use super::tree::TreeNode;
use std::rc::Rc;
use std::cell::RefCell;
use std::cmp::max;

pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    match root {
        Some(n) => {
            let left = n.borrow().left.clone();
            let right = n.borrow().right.clone();
            max(max_depth(left), max_depth(right)) + 1
        },
        None => 0
    }
}

#[test]
fn test_max_depth() {
    let t1 = Some(Rc::new(RefCell::new(TreeNode::new(0))));
    assert_eq!(max_depth(t1),1);
}
